### ISSUE TITLE
Custom Tasks (for Demo-Content install)

### ISSUE THREAD

**Issue by mowshon** [ThemeFuse/Unyson#2126]
In some case we need to use own tasks in demo-install process. In current `Unyson-Backups-Extension` version, we can't add new task-type and we can't put new task in install queue.

_Tasks Registration:_
https://github.com/ThemeFuse/Unyson-Backups-Extension/blob/master/includes/module/tasks/type/init.php

_Demo-Content install proccess:_
https://github.com/ThemeFuse/Unyson-Backups-Extension/blob/master/extensions/backups-demo/class-fw-extension-backups-demo.php#L390-L410

**_Why we need this?_**
Some demo-content can add new task to force-install new plugin or unyson-extension in background.

**Comment by danyj** [comment]
@mowshon , if is added , can you please show an example?

**Comment by mowshon** [comment]
In my first post I made a mistake, we **CAN** register new task using `fw_ext_backups_task_types_register` action.

https://github.com/ThemeFuse/Unyson-Backups-Extension/blob/master/includes/module/tasks/type/init.php#L41

In our functions.php
```
function _action_register_backups_task_types( $task_types ) {
	require_once '/home/our-theme/tasks/class-fw-ext-backups-task-type-test.php';
	$task_types->register(new FW_Ext_Backups_Task_Type_Test());
}

add_action(
	'fw_ext_backups_task_types_register',
	'_action_register_backups_task_types'
);
```

The last problem is to put our task in install queue.

**Comment by mowshon** [comment]
@danyj you can look at https://github.com/ThemeFuse/Unyson-Backups-Extension/blob/master/extensions/backups-demo/class-fw-extension-backups-demo.php#L390-L410

In this area you can add new task.

**Custom task**. (How to register new task I show in last comment)
```php
class FW_Ext_Backups_Task_Type_Some extends FW_Ext_Backups_Task_Type {
	public function get_type() {
		return 'some';
	}

	public function get_title(array $args = array(), array $state = array()) {
		return __('Run Some Task', 'fw');
	}

	public function execute(array $args, array $state = array()) {
		sleep(5);
                // do your stuff, install plugins, create folders, etc.
		return true;
	}
}
```

in `class-fw-extension-backups-demo.php`
https://github.com/ThemeFuse/Unyson-Backups-Extension/blob/master/extensions/backups-demo/class-fw-extension-backups-demo.php#L390-L410

```php
$collection->add_task(new FW_Ext_Backups_Task(
    $id_prefix .'tester',
    'some',  // key from get_type()
    array()
));
```

Result:
![Example](https://i.gyazo.com/71e724d47039f115a298da6c3fa681c0.png)

**Comment by moldcraft** [comment]
> force-install new plugin or unyson-extension in background

You can't do that on demo install, because on some servers the user is required to enter the FTP credentials on plugins and extensions install, because the install [requires `WP_Filesystem`](https://github.com/ThemeFuse/Unyson/blob/962a9591904e3015e0fe19cdf30271065f597213/framework/core/components/extensions/manager/class--fw-extensions-manager.php#L1010) to be initialized.

**Comment by danyj** [comment]
>You can't do that 

@moldcraft  why do you always have to kill the fun :)  just kidding , honestly , we know that those are not Unyson limits .

**Comment by moldcraft** [comment]
> why do you always have to kill the fun

Because I had cases when some developers have implemented very quick some featured requested by users or by our boss/manager, without thinking very good about the implications/consequences/extendability, then we had hard times fixing it ([for example](https://github.com/ThemeFuse/Unyson/issues/1722#issuecomment-230566810))

**Comment by danyj** [comment]
Cmon bro , I said I was kidding ,  and I mean it.  You know we all LOVEEEEEEEEE @moldcraft  :)

**Comment by danyj** [comment]
![1cjfux](https://cloud.githubusercontent.com/assets/2340172/19475802/3445ffaa-9536-11e6-960f-5cc8af557e74.jpg)

**Comment by mowshon** [comment]
@moldcraft, yes, when `FS_METHOD` is not `direct` we can't install plugins and user will be informed to install plugins manually.
```php
define('FS_METHOD', 'ftpext');
```

Usually, FS_METHOD allow to install directly.

**Comment by moldcraft** [comment]
I have added 3 actions but in your case I think you should use [this](https://github.com/ThemeFuse/Unyson-Backups-Extension/commit/313a2767f248c20cdb8c6f639a420da74a839bdf#diff-5c1676521439608717dba336a3bfc0e5R413)
